### DISCUSSION TITLE
Proposal: allow asterisk (*) in method calls to use the default

### DISCUSSION THREAD

**Discussion by sliekens** [dotnet/csharplang#7640] [Language Ideas] [answered]
Hi,

This proposal is about making it easier to pass values for optional arguments in method calls.

To clarify, if you have a method with multiple optional arguments, and you only need to specify a value for some of them, you need to resort to named arguments instead of positional arguments.

Take this example:

``` csharp
void GitCommit(string message, bool amend = false, bool allowEmpty = false, bool dryRun = false)
{
}
```

Today, you need to use named arguments a lot to work with these kinds of methods.

``` csharp
GitCommit("required", dryRun: true);
GitCommit("required", allowEmpty: true);
GitCommit("required", true, dryRun: true);
```

Often I would prefer to use positional arguments but there is nothing in the language today that makes it possible. With this proposal I would like to introduce the asterisk (*) operator to make it possible to call such methods using positional arguments:

``` csharp
GitCommit("required", *, *, true);
GitCommit("required", *, true);
GitCommit("required", true, *, true);
```

The asterisk would indicate that the default value for the argument should be used, which could be anything (hence the asterisk.) So it would be lowered into code which you would otherwise have to write yourself:

``` csharp
// lowered code
GitCommit("required", false, false, true);
GitCommit("required", false, true);
GitCommit("required", true, false, true);
```

This example is of course trivial for demonstration purpose. In the real world, these methods can get really hairy, especially when optional strings with a default value are involved.

Please also consider this: we often already have a named variable that we pass on to the method call. For example this could be a GUI for Git where we bind checkboxes to variables.

``` csharp
var message = MessageControl.Text;
var amend = AmendControl.Checked;
var dryRun = DryRunControl.Checked;

GitCommit(message, amend, dryRun: dryRun);
```

Could instead be rewritten as:

``` csharp
var message = MessageControl.Text;
var amend = AmendControl.Checked;
var dryRun = DryRunControl.Checked;

GitCommit(message, amend, *, dryRun);
```

I also expect some people might use this to make it **unambiguously clear** that default values are at play:

``` csharp
GitCommit(message, *, *, *);
```

Or even more explicitly:

``` csharp
GitCommit(message, amend: *, allowEmpty: *, dryRun: *);
```

Although enabling that style of programming is not a primary goal of my request.

Thank you for your time and consideration.

**Comment by orthoxerox** [comment]
```c#
GitCommit("required", default, default, true);
GitCommit("required", default, true);
GitCommit("required", true, default, true);
```

**Reply by yaakov-h** [reply]
This works for all cases supplied in the OP but wouldn't work for `bool someVar = true`, or "optional strings with a default value", assuming the OP is referring to non-null default values.

**Reply by sliekens** [reply]
Yeah this is not a solution. The `default` keyword doesn't use the default value for an argument, it uses the default value for a type.

**Comment by DavidArno** [comment]
The art of good code is readability. Easy to read code is easier to understand and so is less prone to mistakes, and is quicker and easier to modify and support. With that in mind, I am utterly opposed to this  as `GitCommit("required", *, *, true);` is the opposite of easy to read code. Boolean parameters are bad enough at the best of times, but at least the current default parameter implementation encourages the use of named parameters. It's not perfect though as your `GitCommit("required", true, dryRun: true);` example shows: I still need to check the method signature to determine what that first `true` does.

For me, Python's "keyword-only parameters" approach is far more desirable, where we'd specify the method like this:
```cs
void GitCommit(string message, *, bool amend = false, bool allowEmpty = false, bool dryRun = false) ...
```
with the `*` parameter making naming the following parameters mandatory. 

In lieu of that, the best practice in C# is to define a flags enum,
```cs
[Flags]
enum GitOptions
{
    None = 0b000,
    Amend = 0b001,
    AllowEmpty = 0b010,
    DryRun = 0b100
}

void GitCommit(string message, GitOptions options = GitOptions.None) ... 
```
Your examples become far more readable then:
```cs
GitCommit("required");
GitCommit("required", GitOptions.DryRun);
GitCommit("required", GitOptions.AllowEmpty);
GitCommit("required", GitOptions.Amend|GitOptions.DryRun);
```

**Reply by sliekens** [reply]
I'm with you on this but I have two thoughts:

Positional arguments at least remind the reader that there are more optional arguments than the ones supplied by named arguments. Like you said, you have to check the method signature no matter which style you prefer.

Regarding good method design, we don't always get to choose the methods we have to use. So this is really about enabling the use of positional arguments if you have no control over the method signature (or if changing it would be too much of a hassle.) 

---

If that was not convincing, please also consider this: we often already have a named variable that we pass on to the method call. For example this could be a GUI for Git where we bind checkboxes to variables.

``` csharp
var message = MessageControl.Text;
var amend = AmendControl.Checked;
var dryRun = DryRunControl.Checked;

GitCommit(message, amend, dryRun: dryRun);
```

Could instead be rewritten as:


``` csharp
var message = MessageControl.Text;
var amend = AmendControl.Checked;
var dryRun = DryRunControl.Checked;

GitCommit(message, amend, *, dryRun);
```

**Reply by iam3yal** [reply]
@sliekens But then if you can't change existing apis you can always create wrappers around them so they would be more user-friendly.

**Reply by sliekens** [reply]
Sure but this is getting off-topic. This is not about API design, this is about adding another way to consume existing APIs with optional arguments.

**Reply by theunrepentantgeek** [reply]
It's not off topic; it's at worst topic-adjacent.

The assertion being made is that C# needs another way to handle optional parameters because the techniques currently available are dramatically insufficient. The prevalence of this kind of API is relevant - as is the question of whether C# should encourage APIs with this design.

If there aren't many APIs that take this approach, then the return-on-investment for the change is negative.

**Reply by sliekens** [reply]
I am sure there are beautiful APIs with optional arguments, my example is just not one of them. In my opinion, not having a way to specify a default for an argument is simply an oversight in the C# language.

For example JavaScript has `undefined` to enable this.

``` js
function GitCommit(message, amend = false, allowEmpty = false, dryRun = false) {
    let cmd = `git commit -m '${message}'`
        + (amend ? ' --amend' : '')
        + (allowEmpty ? ' --allow-empty' : '')
        + (dryRun ? ' --dry-run' : '');
    console.log(cmd);
}

GitCommit('required', undefined, undefined, true);
```

Outputs:
> git commit -m 'required'   --dry-run

It's almost weird to me that C# has nothing similar.

**Reply by HaloFour** [reply]
One feeds into the other.  The API design is influenced a great on idiomatic use of the language.  When one designs APIs in one language based on the idioms of another language they are guaranteed to run into friction.  The language isn't intended to support all idioms.

In this case specifically, the idiomatic approach would be the use of a flag enum, or named arguments.  Wanting to use positional arguments here would be frowned upon, thus I don't see why the language would encourage it through specifically supporting it.  Such a feature would make the code much harder to read since you need to memorize what some combination of arbitrary Boolean values in arbitrary positions means.

**Reply by sliekens** [reply]
Yeah, I really disagree that flags or named arguments are **always** better which is why I wanted to talk about this. It's often the case that you already have some named variables, like in my other comment.

I find the named arguments approach with named variables a bit awkward:

``` csharp
var message = MessageControl.Text;
var allowEmpty = allowEmptyControl.Checked;
var dryRun = DryRunControl.Checked;

GitCommit(message, allowEmpty: allowEmpty, dryRun: dryRun);
```

Same with flags, I don't think this is great code:

``` csharp
var message = MessageControl.Text;
var allowEmpty = allowEmptyControl.Checked;
var dryRun = DryRunControl.Checked;

var options = GitOptions.None;
if (allowEmpty) options |= GitOptions.AllowEmpty;
if (dryRun) options |= GitOptions.DryRun;

GitCommit(message, options);
```

I would much prefer this:


``` csharp
var message = MessageControl.Text;
var allowEmpty = allowEmptyControl.Checked;
var dryRun = DryRunControl.Checked;

GitCommit(message, *, allowEmpty, dryRun);
```

**Reply by HaloFour** [reply]
You can prefer it all you want, that doesn't mean that C# needs to go out of it's way to support it over what it considers to be more idiomatic patterns.

**Reply by sliekens** [reply]
Why do you consider it non-idiomatic? Would it help if it was a keyword instead of a special character? Otherwise, it's just a method call. I'm not proposing wild new ideas.

``` c#
GitCommit(message, defaultvalue, allowEmpty, dryRun);
```

**Reply by HaloFour** [reply]
Microsoft considers it non-idiomatic, which is why the .NET Framework guidelines explicitly state to [favor method overloading instead of default values](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/member-overloading), and to [use enums instead of having multiple Boolean parameters](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/parameter-design#choosing-between-enum-and-boolean-parameters).

> Would it help if it was a keyword instead of a special character?

No, it's just as unreadable because nothing explains what any of those parameters are.

**Reply by sliekens** [reply]
Okay that's good to know but I would like to underscore that those are framework/library design guidelines, you're still free to use optional arguments as you see fit in application code. So with that, I don't think it's fair to call optional arguments non-idiomatic.

> No, it's just as unreadable because nothing explains what any of those parameters are.

That's a bit the point of the `*`, like in ~~regular expressions~~ glob patterns, it can be whatever and you're only interested in what comes before or after.

In fact `*` enables something that's not possible today: explicitly using the default value using a named argument:

``` csharp
GitCommit(message, amend: *, allowEmpty, dryRun);
```

**Reply by CyrusNajmabadi** [reply, marked as answer]
Fundamentally, what the ecosystem does is "idiomatic". And the ecosystem tends to follow the design guidelines to not rock the boat and to make .net feel like a consistent and complimentary set of tools and libraries. We're unlikely to change the language for the niche case. Unless we truly believed it was better and there was enough value in dragging *everyone* to that new place. 

That's a tall order, and nothing so far in this space has given us the motivation to go that route.
